File: backend/templates/go/internal/middlewares/request-context.middleware.go
package middleware

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"template/1.0/global"

	"github.com/gin-gonic/gin"
)

func RequestContextMiddleware(c *gin.Context) {
	tokenPrefix := "Bearer "

	cid := c.GetHeader(global.CID_HEADER_KEY)
	accessToken := strings.TrimPrefix(c.GetHeader(global.ACCESS_TOKEN_HEADER_KEY), tokenPrefix)
	requestTimestamp := time.Now().UnixMilli()

	ctx := global.Ctx{
		Cid:              cid,
		RequestTimestamp: requestTimestamp,
		AccessToken:      accessToken,
	}

	userInfoHeader := c.GetHeader(global.USER_INFO_HEADER_KEY)
	if userInfoHeader != "" {
		userInfoBytes, err := base64.StdEncoding.DecodeString(userInfoHeader)
		if err == nil {
			if err := json.Unmarshal(userInfoBytes, &ctx.UserInfo); err != nil {
				fmt.Println("Error parsing user info:", err)
			}
		}
	}

	requestInfoHeader := c.GetHeader(global.REQUEST_INFO_HEADER_KEY)
	if requestInfoHeader != "" {
		requestInfoBytes, err := base64.StdEncoding.DecodeString(requestInfoHeader)
		if err == nil {
			if err := json.Unmarshal(requestInfoBytes, &ctx.RequestInfo); err != nil {
				fmt.Println("Error parsing request info:", err)
			}
		}
	}

	c.Set("Ctx", ctx)
	c.Next()
}

File: backend/node-common/src/middlewares/request-context.middleware.ts
import { AUTHORIZATION, CID_HEADER_KEY, USER_INFO_HEADER_KEY } from '~/constants';
import { Injectable, NestMiddleware } from '@nestjs/common';
// import { RequestContext } from '~/models';
// import { generateCID } from '~/utils';
import { FastifyRequest, FastifyReply } from 'fastify';
// import { AslRequestContext } from '~/shared';
/**
 * TODO not working with Fastify
 */
@Injectable()
export class RequestContextMiddleware implements NestMiddleware {
  constructor() {}
  use(request: FastifyRequest['raw'], _: FastifyReply['raw'], next: () => void) {
    // const tokenPrefix = 'Bearer ';
    // const cid =
    //   (request.headers[CID_HEADER_KEY] as string) ||
    //   (request.headers[CID_HEADER_KEY.toUpperCase()] as string);
    // const authorization =
    //   (request.headers[AUTHORIZATION] as string) ||
    //   (request.headers[AUTHORIZATION.toUpperCase()] as string);
    // request.b = new RequestContext({
    //   cid: generateCID(cid),
    //   requestTimestamp: this.getTimestamp(),
    //   accessToken: authorization?.replace(tokenPrefix, ''),
    // });
    // const userInfoBase64 = request.headers[USER_INFO_HEADER_KEY] as string;
    // if (userInfoBase64) {
    //   try {
    //     const userInfoStr = Buffer.from(userInfoBase64, 'base64').toString('utf-8');
    //     request.ctx.userInfo = JSON.parse(userInfoStr);
    //   } catch (error) {
    //     console.warn('Error parsing user info', error);
    //   }
    // }
    // AslRequestContext.cls.run(new AslRequestContext(request, _), next);
  }

  getTimestamp = () => new Date().getTime();
}
